### ISSUE TITLE
Closures and channel operations result in a bug

### ISSUE THREAD

**Issue by aykevl** [tinygo-org/tinygo#508]
The following doesn't produce the expected output:

```go
package main

var strch = make(chan string)

func main() {
	strch := strch // comment this line to avoid the bug
	go func() {
		println("sending...")
		strch <- "hello"
		println("sent!")
	}()
	println(<-strch)
}
```

There appears to be some sort of interaction between closures and channel operations that goes wrong.

**Comment by jadr2ddude** [comment]
It appears that we are accidentally passing in `undef` for the context. As a result, the closure ends up with a channel pointer that actually points to something else (when tested against #513, I ended up with `ch=0x401802 <runtime.fakeCoroutine>`).

**Comment by jadr2ddude** [comment]
And it does this because it is treated as a static call, and therefore we never actually do anything with the closure context.

**Comment by jadr2ddude** [comment]
As it turns out, this fails to compile with assertions enabled.

**Comment by aykevl** [comment]
Fixed by #640.
